File: backend/src/main/java/com/poeflipfinder/backend/usecase/computeflipopportunities/BulkBuyOpportunityFinder.java
package com.poeflipfinder.backend.usecase.computeflipopportunities;

import com.poeflipfinder.backend.entity.Currency;
import com.poeflipfinder.backend.entity.CurrencyAmount;
import com.poeflipfinder.backend.entity.ExchangeMarketSnapshot;
import com.poeflipfinder.backend.entity.FlipOpportunity;
import com.poeflipfinder.backend.entity.Technique;
import com.poeflipfinder.backend.entity.UndercutQuote;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * Computes Feature E (Bulk Buy / triangular arbitrage, docs/PRD.md § 7.5)
 * opportunities: for any currency X that trades against both Chaos and
 * Divine, checks whether buying X with one and reselling it for the other
 * beats the direct Chaos<->Divine exchange rate -- evaluated in both
 * directions independently, since undercut pricing isn't symmetric (one
 * direction can be viable while the reverse isn't).
 */
class BulkBuyOpportunityFinder {

    List<FlipOpportunity> find(List<ExchangeMarketSnapshot> snapshots, DivineChaosRate rate) {
        if (rate == null) {
            return List.of();
        }

        Map<Currency, ExchangeMarketSnapshot> chaosLegs =
                legsByOtherCurrency(snapshots, BaseCurrencyIds.CHAOS_EXTERNAL_ID);
        Map<Currency, ExchangeMarketSnapshot> divineLegs =
                legsByOtherCurrency(snapshots, BaseCurrencyIds.DIVINE_EXTERNAL_ID);

        List<FlipOpportunity> opportunities = new ArrayList<>();
        for (Map.Entry<Currency, ExchangeMarketSnapshot> entry : chaosLegs.entrySet()) {
            Currency intermediary = entry.getKey();
            ExchangeMarketSnapshot divineSnapshot = divineLegs.get(intermediary);
            if (divineSnapshot == null) {
                continue;
            }

            Optional<UndercutQuote> chaosQuote = quoteFor(entry.getValue(), BaseCurrencyIds.CHAOS_EXTERNAL_ID);
            Optional<UndercutQuote> divineQuote = quoteFor(divineSnapshot, BaseCurrencyIds.DIVINE_EXTERNAL_ID);
            if (chaosQuote.isEmpty() || divineQuote.isEmpty()) {
                continue;
            }

            addIfPresent(opportunities, chaosToDivine(intermediary, chaosQuote.get(), divineQuote.get(), rate));
            addIfPresent(opportunities, divineToChaos(intermediary, chaosQuote.get(), divineQuote.get(), rate));
        }
        return opportunities;
    }

    /**
     * All snapshots where one side is the given base currency, keyed by the
     * other ("intermediary") side -- excluding the base-to-base reference
     * pair itself (e.g. Chaos<->Divine), which isn't a valid intermediary.
     */
    private Map<Currency, ExchangeMarketSnapshot> legsByOtherCurrency(
            List<ExchangeMarketSnapshot> snapshots, String baseExternalId) {
        Map<Currency, ExchangeMarketSnapshot> legs = new HashMap<>();
        for (ExchangeMarketSnapshot snapshot : snapshots) {
            Currency other = otherCurrency(snapshot, baseExternalId);
            if (other == null || isBaseCurrency(other)) {
                continue;
            }
            legs.put(other, snapshot);
        }
        return legs;
    }

    private Currency otherCurrency(ExchangeMarketSnapshot snapshot, String baseExternalId) {
        if (baseExternalId.equals(snapshot.currencyA().externalId())) {
            return snapshot.currencyB();
        }
        if (baseExternalId.equals(snapshot.currencyB().externalId())) {
            return snapshot.currencyA();
        }
        return null;
    }

    private boolean isBaseCurrency(Currency currency) {
        return BaseCurrencyIds.CHAOS_EXTERNAL_ID.equals(currency.externalId())
                || BaseCurrencyIds.DIVINE_EXTERNAL_ID.equals(currency.externalId());
    }

    private Optional<UndercutQuote> quoteFor(ExchangeMarketSnapshot snapshot, String baseExternalId) {
        boolean baseIsA = baseExternalId.equals(snapshot.currencyA().externalId());
        double lowestRatioStart = baseIsA ? snapshot.lowestRatioA() : snapshot.lowestRatioB();
        double lowestRatioVia = baseIsA ? snapshot.lowestRatioB() : snapshot.lowestRatioA();
        double highestRatioStart = baseIsA ? snapshot.highestRatioA() : snapshot.highestRatioB();
        double highestRatioVia = baseIsA ? snapshot.highestRatioB() : snapshot.highestRatioA();
        long lowestStockStart = baseIsA ? snapshot.lowestStockA() : snapshot.lowestStockB();
        long highestStockStart = baseIsA ? snapshot.highestStockA() : snapshot.highestStockB();
        return UndercutQuote.resolve(
                lowestRatioStart, lowestRatioVia, highestRatioStart, highestRatioVia,
                lowestStockStart, highestStockStart);
    }

    /**
     * Buy the intermediary with Chaos, sell it for Divine, compare against
     * the direct Chaos->Divine baseline at the raw, non-undercut, averaged
     * rate -- a comparison reference, not an order we're suggesting the user
     * post.
     *
     * <p>Scaled so the trade always ends at exactly 1 Divine sold, rather
     * than starting from 1 Chaos: "1 Chaos -> 0.005 Divine" is a meaningless
     * fraction to read, since Chaos is the much smaller-value currency here.
     * "~230 Chaos -> 1 Divine" is how a player actually thinks about this
     * direction, and lines up directly with the "direct ~225 c/div" figure
     * in the detail line. Margin is scale-invariant (a ratio); profit scales
     * with the trade size, same as it would if a player actually multiplied
     * the trade up. Per-unit rates (buy/sell/direct) and volume (real market
     * depth, independent of how much of it we choose to reference) are
     * unaffected by the rescaling either way.
     */
    private FlipOpportunity chaosToDivine(
            Currency intermediary, UndercutQuote chaosLeg, UndercutQuote divineLeg, DivineChaosRate rate) {
        if (chaosLeg.suggestedBuyPrice().isEmpty()) {
            return null;
        }
        double sellAmount = 1.0;
        double viaAmount = divineLeg.suggestedSellPrice(); // sells for exactly 1 Divine, by definition
        double startAmount = viaAmount / chaosLeg.suggestedBuyPrice().get();

        double directBaselineDivine = startAmount / rate.chaosPerDivine();
        double marginPercent = (sellAmount - directBaselineDivine) / directBaselineDivine * 100;
        double profitChaos = (sellAmount - directBaselineDivine) * rate.chaosPerDivine();
        double volume = Math.min(chaosLeg.buyLegStock(), divineLeg.sellLegStock());
        String detail = detail(chaosLeg.suggestedBuyPrice().get(), divineLeg.suggestedSellPrice(), rate.chaosPerDivine());

        return new FlipOpportunity(
                Technique.BULK_BUY,
                List.of(new CurrencyAmount(rate.chaosCurrency(), startAmount)),
                List.of(new CurrencyAmount(intermediary, viaAmount)),
                List.of(new CurrencyAmount(rate.divineCurrency(), sellAmount)),
                marginPercent,
                new CurrencyAmount(rate.chaosCurrency(), profitChaos),
                volume,
                detail);
    }

    /**
     * Buy the intermediary with 1 Divine, sell it for Chaos, compare against
     * the direct Divine->Chaos baseline -- already Chaos-denominated, so the
     * profit figure needs no conversion.
     */
    private FlipOpportunity divineToChaos(
            Currency intermediary, UndercutQuote chaosLeg, UndercutQuote divineLeg, DivineChaosRate rate) {
        if (divineLeg.suggestedBuyPrice().isEmpty()) {
            return null;
        }
        double viaAmount = divineLeg.suggestedBuyPrice().get();
        double sellAmountChaos = viaAmount / chaosLeg.suggestedSellPrice();
        double directBaselineChaos = rate.chaosPerDivine();
        double marginPercent = (sellAmountChaos - directBaselineChaos) / directBaselineChaos * 100;
        double profitChaos = sellAmountChaos - directBaselineChaos;
        double volume = Math.min(divineLeg.buyLegStock(), chaosLeg.sellLegStock());
        String detail = detail(viaAmount, chaosLeg.suggestedSellPrice(), rate.chaosPerDivine());

        return new FlipOpportunity(
                Technique.BULK_BUY,
                List.of(new CurrencyAmount(rate.divineCurrency(), 1.0)),
                List.of(new CurrencyAmount(intermediary, viaAmount)),
                List.of(new CurrencyAmount(rate.chaosCurrency(), sellAmountChaos)),
                marginPercent,
                new CurrencyAmount(rate.chaosCurrency(), profitChaos),
                volume,
                detail);
    }

    private void addIfPresent(List<FlipOpportunity> opportunities, FlipOpportunity opportunity) {
        if (opportunity != null) {
            opportunities.add(opportunity);
        }
    }

    private String detail(double buyPrice, double sellPrice, double chaosPerDivine) {
        return "buy %s:1 · sell %s:1 · direct ~%s c/div"
                .formatted(formatRatio(buyPrice), formatRatio(sellPrice), formatRatio(chaosPerDivine));
    }

    private String formatRatio(double value) {
        if (value == Math.floor(value) && Double.isFinite(value)) {
            return String.valueOf((long) value);
        }
        return "%.2f".formatted(value);
    }
}

File: app/lib/poe_flip_finder/bulk_buy_opportunity_finder.ex
defmodule PoeFlipFinder.BulkBuyOpportunityFinder do
  @moduledoc """
  Computes Feature E (Bulk Buy / triangular arbitrage, docs/PRD.md § 7.5)
  opportunities: for any currency X that trades against both Chaos and
  Divine, checks whether buying X with Divine and reselling it for Chaos
  beats the direct Divine->Chaos exchange rate.

  Only this one direction is computed -- per direct user feedback, the
  reverse (buy with Chaos, resell for Divine) isn't a trade players are
  looking for here, even though it's a real, computable arbitrage. Bulk
  Buy is specifically "convert Divine you're already holding into more
  Chaos than a direct sale would get," not a general two-way triangular
  scan.
  """

  alias PoeFlipFinder.{
    BaseCurrencyIds,
    Currency,
    CurrencyAmount,
    DivineChaosRate,
    ExchangeMarketSnapshot,
    FlipOpportunity,
    UndercutQuote
  }

  @spec find([ExchangeMarketSnapshot.t()], DivineChaosRate.t() | nil) :: [FlipOpportunity.t()]
  def find(_snapshots, nil), do: []

  def find(snapshots, rate) do
    chaos_legs = legs_by_other_currency(snapshots, BaseCurrencyIds.chaos_external_id())
    divine_legs = legs_by_other_currency(snapshots, BaseCurrencyIds.divine_external_id())

    Enum.flat_map(chaos_legs, fn {intermediary, chaos_snapshot} ->
      case Map.fetch(divine_legs, intermediary) do
        :error ->
          []

        {:ok, divine_snapshot} ->
          opportunities_for(intermediary, chaos_snapshot, divine_snapshot, rate)
      end
    end)
  end

  defp opportunities_for(intermediary, chaos_snapshot, divine_snapshot, rate) do
    chaos_quote = quote_for(chaos_snapshot, BaseCurrencyIds.chaos_external_id())
    divine_quote = quote_for(divine_snapshot, BaseCurrencyIds.divine_external_id())

    if chaos_quote == nil or divine_quote == nil do
      []
    else
      case divine_to_chaos(intermediary, chaos_quote, divine_quote, rate) do
        nil -> []
        opportunity -> [opportunity]
      end
    end
  end

  # All snapshots where one side is the given base currency, keyed by the
  # other ("intermediary") side -- excluding the base-to-base reference
  # pair itself (e.g. Chaos<->Divine), which isn't a valid intermediary.
  @spec legs_by_other_currency([ExchangeMarketSnapshot.t()], String.t()) ::
          %{Currency.t() => ExchangeMarketSnapshot.t()}
  defp legs_by_other_currency(snapshots, base_external_id) do
    Enum.reduce(snapshots, %{}, fn snapshot, legs ->
      add_leg(legs, other_currency(snapshot, base_external_id), snapshot)
    end)
  end

  defp add_leg(legs, nil, _snapshot), do: legs

  defp add_leg(legs, other, snapshot) do
    if base_currency?(other), do: legs, else: Map.put(legs, other, snapshot)
  end

  defp other_currency(snapshot, base_external_id) do
    cond do
      snapshot.currency_a.external_id == base_external_id -> snapshot.currency_b
      snapshot.currency_b.external_id == base_external_id -> snapshot.currency_a
      true -> nil
    end
  end

  defp base_currency?(currency) do
    currency.external_id in [
      BaseCurrencyIds.chaos_external_id(),
      BaseCurrencyIds.divine_external_id()
    ]
  end

  defp quote_for(snapshot, base_external_id) do
    base_is_a? = snapshot.currency_a.external_id == base_external_id

    UndercutQuote.resolve(
      pick(base_is_a?, snapshot.lowest_ratio_a, snapshot.lowest_ratio_b),
      pick(base_is_a?, snapshot.lowest_ratio_b, snapshot.lowest_ratio_a),
      pick(base_is_a?, snapshot.highest_ratio_a, snapshot.highest_ratio_b),
      pick(base_is_a?, snapshot.highest_ratio_b, snapshot.highest_ratio_a),
      pick(base_is_a?, snapshot.lowest_stock_a, snapshot.lowest_stock_b),
      pick(base_is_a?, snapshot.highest_stock_a, snapshot.highest_stock_b),
      pick(base_is_a?, snapshot.volume_traded_a, snapshot.volume_traded_b),
      pick(base_is_a?, snapshot.volume_traded_b, snapshot.volume_traded_a)
    )
  end

  defp pick(true, a, _b), do: a
  defp pick(false, _a, b), do: b

  # Buy the intermediary with 1 Divine, sell it for Chaos, compare against
  # the direct Divine->Chaos baseline -- already Chaos-denominated, so the
  # profit figure needs no conversion.
  #
  # The buy leg uses suggested_buy_price (a competitive limit order,
  # undercut -1) but the sell leg uses market_sell_price -- see
  # UndercutQuote for why that is the hour's worse-for-a-seller extreme, not
  # suggested_sell_price's round-trip-favorable one. Posting two unfilled
  # limit orders back to back on a large, often-illiquid intermediary stack
  # is also a bigger risk than posting one, so only the entry is priced
  # competitively; the exit assumes dumping into demand that's already been
  # shown to exist, at worst at the less generous of the hour's two real
  # rates.
  defp divine_to_chaos(
         _intermediary,
         _chaos_quote,
         %UndercutQuote{suggested_buy_price: nil},
         _rate
       ),
       do: nil

  defp divine_to_chaos(
         _intermediary,
         %UndercutQuote{market_sell_price: price},
         _divine_quote,
         _rate
       )
       when price <= 0,
       do: nil

  defp divine_to_chaos(intermediary, chaos_quote, divine_quote, rate) do
    via_amount = divine_quote.suggested_buy_price
    sell_amount_chaos = via_amount / chaos_quote.market_sell_price
    direct_baseline_chaos = rate.chaos_per_divine
    margin_percent = (sell_amount_chaos - direct_baseline_chaos) / direct_baseline_chaos * 100
    profit_chaos = sell_amount_chaos - direct_baseline_chaos
    volume = min(divine_quote.buy_leg_stock, chaos_quote.buy_leg_stock)

    %FlipOpportunity{
      technique: :bulk_buy,
      start: [%CurrencyAmount{currency: rate.divine_currency, quantity: 1.0}],
      via: [%CurrencyAmount{currency: intermediary, quantity: via_amount}],
      sell: [%CurrencyAmount{currency: rate.chaos_currency, quantity: sell_amount_chaos}],
      margin_percent: margin_percent,
      profit: %CurrencyAmount{currency: rate.chaos_currency, quantity: profit_chaos},
      start_chaos_equivalent: DivineChaosRate.to_chaos(rate, rate.divine_currency, 1.0),
      volume: volume,
      detail: detail(via_amount, chaos_quote.market_sell_price, rate.chaos_per_divine)
    }
  end

  defp detail(buy_price, sell_price, chaos_per_divine) do
    "buy #{format_ratio(buy_price)}:1 · sell #{format_ratio(sell_price)}:1 · direct ~#{format_ratio(chaos_per_divine)} c/div"
  end

  defp format_ratio(value) do
    if Float.floor(value) == value do
      value |> trunc() |> Integer.to_string()
    else
      :erlang.float_to_binary(value, decimals: 2)
    end
  end
end
